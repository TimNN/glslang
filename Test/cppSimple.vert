#version 400

#define ON

float sum = 0.0;

void main()
{

#ifdef ON
//yes
sum += 1.0;
#endif

#ifdef OFF
//no
sum += 20.0;
#endif

#if defined(ON)
//yes
sum += 300.0;
#endif

#if defined(OFF)
//no
sum += 4000.0;
#endif

#if !defined(ON)
//no
sum += 50000.0;
#endif

#ifndef OFF
//yes
sum += 600000.0;
#else
//no
sum += 0.6;
#endif

#if defined(ON) && defined(OFF)
//no
sum += 0.7;
#elif !defined(OFF)
//yes
sum += 7000000.0;
#endif

#if defined(ON) && !defined(OFF)
//yes
sum += 80000000.0;
#endif

#if defined(OFF) || defined(ON)
//yes
sum += 900000000.0;
#endif

#if NEVER_DEFINED
//no
sum += 0.04;
#else
sum += 0.05;
#endif

// sum should be 987600301.7
    gl_Position = vec4(sum);
}

#define A 0
#define B 0
#define C 0

#if (A == B) || (A == C)
#error good1
#endif

#if A == B || (A == C)
#error good2
#endif

#if (A == B || (A == C))
#error good3
#endif

#if (AA == BB) || (AA == CC)
#error good4
#endif

#if AA == BB || (AA == CC)
#error good5
#endif

#if ((AA == BB || (AA == CC)))
#error good6
#endif

#if (A == B || (A == C)
#error bad1
#endif

#if A == B || A == C)
#error bad2
#endif

#if (A == B || (A == C)
#error bad3
#endif

#if AA == BB) || (AA == CC)
#error bad4
#endif

#if AA == BB || (AA == CC
#error bad5
#endif

#if ((AA == BB || (AA == CC))))
#error bad6
#endif extra tokens

int linenumber = __LINE__;
int filenumber = __FILE__;
int version = __VERSION__;

#define PI (3.14)
#define TWOPI (2.0 * PI)
float twoPi = TWOPI;

//#define PASTE(a,b) a ## b
//float PASTE(tod, ay) = 17;

"boo" // ERROR
int a = length("aoenatuh");  // ERROR

'int';  // ERROR

// ERROR: all the following are reserved
#define GL_
#define GL_Macro 1
#define __M 
#define M__
#define ABC__DE abc

#if 4
#else extra
#elif
// ERROR elif after else
#endif

#if blah
  #if 0
  #else extra
    #ifdef M
    #else
    #else
    // ERROR else after else
    #endif  extra
  #endif
#endif

#define m1(a,a)  // ERROR
#define m2(a,b)

// okay
#define m3 (a)
#define m3 (a)

// ERROR
#define m4(b)
#define m4 (b)

// ERROR
#define m5 (b)
#define m5(b)

// ERROR
#define m6(a)
#define m6

// ERROR (whitespace)
#define m7 (a)
#define m7 ( a)

#define m80(a,b) is + exactly m3 the same
#define m80(a,b) is + exactly m3 the same

// ERROR
#define m8(a,b) almost + exactly m3 the same
#define m8(a,b) almost + exactly m3 thee same

// ERROR
#define m9(a,b,c) aoe
#define m9(a,d,c) aoe

#define n1 0xf
int n = n1;

#define f1 .08e-2Lf
double f = f1;

#if 1
#else
// ERROR, missing #endif