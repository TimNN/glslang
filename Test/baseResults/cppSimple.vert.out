ERROR: 0:77: '#error' : good1  
ERROR: 0:81: '#error' : good2  
ERROR: 0:85: '#error' : good3  
ERROR: 0:89: '#error' : good4  
ERROR: 0:93: '#error' : good5  
ERROR: 0:97: '#error' : good6  
ERROR: 0:101: 'preprocessor' : expected ')' 
ERROR: 0:101: '#error' : bad1  
WARNING: 0:104: '#if' : unexpected tokens following #if directive - expected a newline 
ERROR: 0:105: '#error' : bad2  
ERROR: 0:109: 'preprocessor' : expected ')' 
ERROR: 0:109: '#error' : bad3  
WARNING: 0:112: '#if' : unexpected tokens following #if directive - expected a newline 
ERROR: 0:113: '#error' : bad4  
ERROR: 0:117: 'preprocessor' : expected ')' 
ERROR: 0:117: '#error' : bad5  
WARNING: 0:120: '#if' : unexpected tokens following #if directive - expected a newline 
ERROR: 0:121: '#error' : bad6  
ERROR: 0:133: '#' : preprocessor directive cannot be preceded by another token 
INTERNAL ERROR: 0:133: Unknown PP token
ERROR: 0:133: '' :  syntax error
ERROR: 17 compilation errors.  No code generated.

ERROR: node is still EOpNull!
0:5  Sequence
0:5    move second child to first child (float)
0:5      'sum' (float)
0:5      0.000000
0:7  Function Definition: main( (void)
0:7    Function Parameters: 
0:12    Sequence
0:12      add second child into first child (float)
0:12        'sum' (float)
0:12        1.000000
0:22      add second child into first child (float)
0:22        'sum' (float)
0:22        300.000000
0:37      add second child into first child (float)
0:37        'sum' (float)
0:37        600000.000000
0:48      add second child into first child (float)
0:48        'sum' (float)
0:48        7000000.000000
0:53      add second child into first child (float)
0:53        'sum' (float)
0:53        80000000.000000
0:58      add second child into first child (float)
0:58        'sum' (float)
0:58        900000000.000000
0:65      add second child into first child (float)
0:65        'sum' (float)
0:65        0.050000
0:69      move second child to first child (4-component vector of float)
0:69        'gl_Position' (gl_Position 4-component vector of float)
0:69        Construct vec4 (4-component vector of float)
0:69          'sum' (float)
0:124  Sequence
0:124    move second child to first child (int)
0:124      'linenumber' (int)
0:124      124 (const int)
0:125  Sequence
0:125    move second child to first child (int)
0:125      'filenumber' (int)
0:125      0 (const int)
0:126  Sequence
0:126    move second child to first child (int)
0:126      'version' (int)
0:126      110 (const int)
0:130  Sequence
0:130    move second child to first child (float)
0:130      'twoPi' (float)
0:130      6.280000

